arwen/contexts/runtime.go

```diff
@@ -833,6 +846,69 @@ func (context *runtimeContext) GetAllErrors() error {
 	return context.errors
 }
 
+// EnableGasTrace sets true to the flag variable that traces gas consumption
+func (context *runtimeContext) EnableGasTrace() {
+	context.currentGasTracer.traceGasEnabled = true
+}
+
+// DisableGasTrace sets true to the flag variable that traces gas consumption
+func (context *runtimeContext) DisableGasTrace() {
+	context.currentGasTracer.traceGasEnabled = false
+}
+
+// TraceGasUsed adds the usedGas passed to the traced gas map with the key as the passed functionName
+func (context *runtimeContext) TraceGasUsed(functionName string, initialGasLeft uint64) {
+	if context.currentGasTracer.traceGasEnabled {
+		scAddress := string(context.scAddress)
+		context.createGasTraceIfNil(scAddress, functionName)
+		usedGas := context.computeUsedGas(initialGasLeft)
+		context.currentGasTracer.gasTrace[scAddress][functionName] = append(context.currentGasTracer.gasTrace[scAddress][functionName], usedGas)
+	}
+}
+
+func (context *runtimeContext) computeUsedGas(initialGasLeft uint64) uint64 {
+	gasLeft := context.host.Metering().GasLeft()
+	return initialGasLeft - gasLeft
+}
+
+// SetInitialGasInGasTrace sets the initial gas in the gasTrace map
+func (context *runtimeContext) SetInitialGasInGasTrace(functionName string) {
+	if context.currentGasTracer.traceGasEnabled && context.currentGasTracer.currentFunctionNameTraced == "" {
+		context.currentGasTracer.currentFunctionNameTraced = functionName
+		scAddress := string(context.scAddress)
+		context.createGasTraceIfNil(scAddress, functionName)
+		context.currentGasTracer.gasTrace[scAddress][functionName] = append(context.currentGasTracer.gasTrace[scAddress][functionName], context.host.Metering().GasLeft())
+	}
+}
+
+func (context *runtimeContext) createGasTraceIfNil(scAddress string, functionName string) {
+	if context.currentGasTracer.gasTrace[scAddress] == nil {
+		context.currentGasTracer.gasTrace[scAddress] = make(map[string][]uint64)
+	}
+	if context.currentGasTracer.gasTrace[scAddress][functionName] == nil {
+		context.currentGasTracer.gasTrace[scAddress][functionName] = make([]uint64, 0)
+	}
+}
+
+// ComputeAndSetUsedGasInGasTrace computes the used gas in the gasTrace map. This function should be used afte initial gas has been set in the gasTrace map using SetInitialGasInGasTrace.
+func (context *runtimeContext) ComputeAndSetUsedGasInGasTrace() {
+	if context.currentGasTracer.traceGasEnabled {
+		functionName := context.currentGasTracer.currentFunctionNameTraced
+		numberOfCalls := len(context.currentGasTracer.gasTrace[string(context.scAddress)][functionName])
```

Comment:
this can pannic if there is nothing under context.scAddress or stuff like that. Need more protection